engines/ic3sa.cpp

```diff
@@ -676,6 +722,116 @@ bool IC3SA::add_to_term_abstraction(const Term & term)
   return new_terms;
 }
 
+void IC3SA::justify_coi(Term c, UnorderedTermSet & projection)
+{
+  // expecting to have a satisfiable context
+  // and IC3Base only solves at context levels > 0
+  assert(solver_context_);
+  to_visit_.clear();
+  visited_.clear();
+  to_visit_.push_back(c);
+
+  TermVec children;
+  UnorderedTermSet free_vars;
+  const UnorderedTermMap & state_updates = ts_.state_updates();
+
+  Term t;
+  while (!to_visit_.empty()) {
+    t = to_visit_.back();
+    to_visit_.pop_back();
+
+    if (visited_.find(t) != visited_.end()) {
+      continue;
+    }
+    visited_.insert(t);
+
+    if (t->get_op() == Ite)
+    {
+      children.clear();
+      children.insert(children.end(), t->begin(), t->end());
+      assert(children.size() == 3);
+      // always visit the condition
+      to_visit_.push_back(children[0]);
+      if (solver_->get_value(children[0]) == solver_true_)
+      {
+        // the if branch is active
+        to_visit_.push_back(children[1]);
+      }
+      else
+      {
+        // the else branch is active
+        to_visit_.push_back(children[2]);
+      }
+    }
+    else if (is_logical_op(t, boolsort_))
+    {
+      Op op = t->get_op();
+      // include bitvector versions for boolector
+      // won't even be in this case if the solver doesn't consider it a boolean
+      // though
+      if ((((op == And) || (op == BVAnd)) && solver_->get_value(t) != solver_true_) ||
+          (((op == Or) || (op == BVOr)) && solver_->get_value(t) == solver_true_))
+      {
+        to_visit_.push_back(get_controlling(t));
+      }
+      else
+      {
+        to_visit_.insert(to_visit_.end(), t->begin(), t->end());
+      }
+    } else if (ts_.is_next_var(t)
+               && state_updates.find(ts_.curr(t)) != state_updates.end()) {
+      to_visit_.push_back(state_updates.at(ts_.curr(t)));
+    } else if (ts_.is_curr_var(t)) {
+      free_vars.insert(t);
+
+      // need to add any constraints that this variable is involved in
+      // to the visit stack
+      for (const auto & elem : constraint_vars_) {
+        if (elem.second.find(t) != elem.second.end()) {
+          // this variable occurs in this constraint
+          // add the constraint
+          to_visit_.push_back(elem.first);
+        }
+      }
+
+    } else {
+      for (const auto & tt : t) {
+        to_visit_.push_back(tt);
+      }
+    }
+  }
+
+  const UnorderedTermSet & inputvars = ts_.inputvars();
```

Comment:
is it used?

Reply:
Nope, good catch! 9c956be8b384ec299f4b614613ffb7af59cdf1fd